src/pj_transform.c

```diff
@@ -81,388 +93,468 @@ static const int transient_error[60] = {
-/*      Do we need to translate from ellipsoidal to geoid vertical      */
-/*      datum?                                                          */
-/* -------------------------------------------------------------------- */
-    if( dstdefn->has_geoid_vgrids && z != NULL )
-    {
-        if( pj_apply_vgridshift( dstdefn, "sgeoidgrids",
-                                 &(dstdefn->vgridlist_geoid),
-                                 &(dstdefn->vgridlist_geoid_count),
-                                 1, point_count, point_offset, x, y, z ) != 0 )
-            return dstdefn->ctx->last_errno;
-    }
 
 /* -------------------------------------------------------------------- */
-/*      But if they are staying lat long, adjust for the prime          */
-/*      meridian if there is one in effect.                             */
+/*            Adjust for the prime meridian if needed.                  */
 /* -------------------------------------------------------------------- */
-    if ((dstdefn->is_geocent || dstdefn->is_latlong) && ( dstdefn->from_greenwich != 0.0 ))
-    {
-        for( i = 0; i < point_count; i++ )
-        {
-            if( x[point_offset*i] != HUGE_VAL )
-                x[point_offset*i] -= dstdefn->from_greenwich;
-        }
+static int prime_meridian (PJ *P, PJ_DIRECTION dir, long n, int dist, double *x) {
+    int i;
+    double pm = P->from_greenwich;
+
+    /* Nothing to do? */
+    if (pm==0.0)
+        return 0;
+    if (!(P->is_geocent || P->is_latlong))
+        return 0;
+
+    if (dir==PJ_FWD)
+        pm = -pm;
+
+    for (i = 0;  i < n;  i++)
+        if (x[dist*i] != HUGE_VAL)
+            x[dist*i] += pm;
+
+    return 0;
 }
 
+
+
 /* -------------------------------------------------------------------- */
-/*      Transform destination latlong to geocentric if required.        */
+/*            Adjust for vertical scale factor if needed                */
 /* -------------------------------------------------------------------- */
-    if( dstdefn->is_geocent )
-    {
-        if( z == NULL )
-        {
-            pj_ctx_set_errno( dstdefn->ctx, PJD_ERR_GEOCENTRIC );
-            return PJD_ERR_GEOCENTRIC;
-        }
+static int height_unit (PJ *P, PJ_DIRECTION dir, long n, int dist, double *z) {
+    int i;
+    double fac = P->vto_meter;
 
-        pj_geodetic_to_geocentric( dstdefn->a_orig, dstdefn->es_orig,
-                                   point_count, point_offset, x, y, z );
+    if (PJ_FWD==dir)
+        fac = P->vfr_meter;
 
-        if( dstdefn->fr_meter != 1.0 )
-        {
-            for( i = 0; i < point_count; i++ )
-            {
-                if( x[point_offset*i] != HUGE_VAL )
-                {
-                    x[point_offset*i] *= dstdefn->fr_meter;
-                    y[point_offset*i] *= dstdefn->fr_meter;
-                    z[point_offset*i] *= srcdefn->fr_meter;
-                }
-            }
-        }
-    }
+    /* Nothing to do? */
+    if (P->vto_meter==0.0)
```

Comment:
Shouldn't this be:
```
if (P->vto_meter==1.0)
```

In 4.9.3 the code was:
```c
/* -------------------------------------------------------------------- */
/*      Transform Z to meters if it isn't already.                      */
/* -------------------------------------------------------------------- */
    if( srcdefn->vto_meter != 1.0 && z != NULL )
    {
        for( i = 0; i < point_count; i++ )
            z[point_offset*i] *= srcdefn->vto_meter;
    }
```

There is also no test for `z != NULL`.

Reply:
@sebastic: That's certainly a blunder - I'll clean this up immediately. Sorry for merging this before seeing your review. I'll open another PR cleaning this up